mod/src/main/kotlin/extracells/core/storage/NBTTagFluidStorage.kt

```diff
@@ -0,0 +1,48 @@
+package extracells.core.storage
+
+import extracells.extension.keys
+import net.minecraft.nbt.NBTTagCompound
+import kotlin.math.min
+
+internal class NBTTagFluidStorage(
+  private val storage: NBTTagCompound,
+  private val maxTypes: Int,
+  private val typeSize: Int,
+  private val size: Int,
+) {
+  val nbtTag: NBTTagCompound
+    get() = this.storage.copy() as NBTTagCompound
+
+  private var freeSpace = size
+
+  /**
+   * @return not injected amount
+   * Example: [amount] = 1000, in [nbtTag] 600, returns will be 400
+   */
+  fun inject(fluidName: String, amount: Int, simulation: Boolean = false): Int {
+    if (freeSpace <= 0) return amount
+
+    val storedFluids = storage.getCompoundTag("fluids")
+    val storedFluid = storedFluids.getInteger(fluidName)
+
+    // if type count is not exceeded
+    return if (storedFluid != 0 || storedFluids.keys.size < this.maxTypes) {
+      val maxAmountToInject = min(freeSpace, amount)
+      storedFluids.setInteger(fluidName, storedFluid + maxAmountToInject)
+      freeSpace -= maxAmountToInject
+      freeSpace -= typeSize
```

Comment:
Why we decrease freeSpace each time on insert fluid? Maybe we want decrease space only if we add new type of fluid?